python/lsst/ts/m2/tcp_client.py

```diff
@@ -197,9 +198,10 @@ async def _put_read_msg_to_queue(self):
             if data is not None:
                 data_decode = data.decode()
                 msg = json.loads(data_decode)
-                self.queue.put_nowait(msg)
-
-                check_queue_size(self.queue, self.log)
+                await self.queue.put(msg)
```

Comment:
The `self.queue.put()` is a blocking call. Therefore, in the unlucky condition that the queue is full, the tcp buffer in kernel might become full or you could not close the application in the blocking condition.

Reply:
According to the [module documentation](https://docs.python.org/3/library/asyncio-queue.html#asyncio.Queue.put):

```
coroutine put(item):
   Put an item into the queue. If the queue is full, wait until a free slot is available before adding the item.
```

So, it is not exactly "blocking" in the  sense that nothing else is going to happen while it waits for a slot.. The component will have a chance the read up the queue while this is waiting to put data. 

You are right though that if the component can't keep up, it probably means that the tcp buffer will start to fill out.. 

But the current code is not giving a chance for the component to read the queue once it is full.

I'll think a bit more about this and do a bit more testing..

Reply:
You could try this to see the code hang up forever:

```
import asyncio                                                          
queue = asyncio.Queue(maxsize=1)                                        
await queue.put(1)                                                      
await queue.put(1) 
```

Reply:
hanging up forever is different than blocking.. That above creates a deadlock because no one is reading the queue..

Reply:
That is because the queue is full. BTW, what I mean is "blocking call".

Reply:
yes I understand.. but blocking is an asyncio context is not exactly blocking the executing.. while the task is waiting to put new stuff on the queue, by calling `await queue.put()` the other tasks have a chance to work on reading the queue and other stuff.. 

But note that when it fails to put, it immediately return. If there's too much data incoming, it might be that the reader queue will never have a chance to read...

I think you re right that my implementation is not the best (as it will invariably cause the tcp buffer to fill up), but I think the original implementation is also problematic... 

I'll think a bit more and to a bit more testing..

Reply:
What I tried to say is in self._monitor_msg(), there is:
```
while self.is_connected():
    await self._put_read_msg_to_queue()
```

If the `await self.queue.put()` blocks in the `self._put_read_msg_to_queue()`, you will have no chance to close the `while` loop in this `asycio` task (of course other tasks can continue to run and check the close condition in `while` loop).

Reply:
This condition would be dealt with by:

https://github.com/lsst-ts/ts_m2/blob/295c461d58e069a85e109f513c522444feae74c2/python/lsst/ts/m2/tcp_client.py#L220

Reply:
If the server closes the connection, the `self.is_connected()` will find it and close the connection. The cancel of task is more like to actively close the connection from the client's side.

Reply:
if the connection is closed, then, most likely the read queue will managed to make enough progress to return the `await self._put_read_msg_to_queue()` call above, which will then, not be called again.

Reply:
I do not think this will happen if the blocking call is used. I guess I had conveyed all the consideration in my mind that should be helpful for you to consider this part of code.

Reply:
I think you are misunderstanding how the `await self.queue.put(...)` works, or maybe I am not understanding what you are trying to convey.

Reply:
For instance, see the module documentation I sent above. The call will wait (not block) only while the queue I full. If there is nothing writing to the queue, the read task will manage to advance and clear it up, and this will immediately make progress as well.

Reply:
What I am considering is always the edge case that the queue is full.

Reply:
ok... that much I understood...

Reply:
The problem with the current implementation is that when the queue fills up, and the traffic continues to be large, there is no allocated time for the reader task to clean it up.... 

I think one possible alternative is to use a rotating queue. Where old messages are discarded once the queue is filled up.

Reply:
I considered this before. However, the `asyncio` does not provide this. We may need to add the `asyncio.lock` with the conventional rotating queue, but it will be resource-intensive. Therefore, I ended up to use the `asyncio.Queue` because it supports the `asyncio` natively.

Reply:
yeah, I don't know.. I think there's something else I must be missing here.. I'll dig a bit further...

Reply:
I mean, the CSC will read this queue to publish to DDS in a `asyncio` task. Therefore, you would need a lock if you are not using the `asyncio.Queue`. I usually try to avoid the use of lock if I could because it is resource-intensive (this means the design of code needs to be careful to avoid the race condition).

Reply:
So, this part here is potentially problematic: https://github.com/lsst-ts/ts_m2/blob/295c461d58e069a85e109f513c522444feae74c2/python/lsst/ts/m2/csc.py#L282-L300

There is no sleep in that loop. So, if there are too many events being published by the controller, the other tasks may never get enough time to process anything..

I think this might be the problem.. I'll try a couple things here and test it.

Reply:
I think you could rewrite it as:

```
        while self._run_loops:

            # Fault the CSC if the controller is in Fault
            if (self.model.controller_state == salobj.State.FAULT) and (
                self.summary_state != salobj.State.FAULT
            ):
                self.fault(
                    code=ErrorCode.ControllerInFault,
                    report="Controller's state is Fault.",
                )

            # Publish the SAL event
            if not self.model.queue_event.empty():
                message = self.model.queue_event.get_nowait()
                self._publish_message_by_sal("evt_", message)

            else:
                await asyncio.sleep(self.timeout_in_second)
```

Reply:
ok.. I looked at it in more details and I think I found a good solution. I retained most of the original implementation, but I added some timers to logging in this method. I also added a report of how many messages the telemetry loop reader is consuming/s.

This should allow us to debug the system more easily. I'm going to run the unit tests a couple more time, touch up a couple things and push it for review.